### DISCUSSION TITLE
Sonos Players

### DISCUSSION THREAD

**Discussion by OzGav** [music-assistant/support#1171] [Player Support]
Documentation https://music-assistant.io/player-support/sonos/

**Comment by habitats-tech** [comment]
I am not certain if anyone has raised this, but Sonos playback is broken (v2.0.0b22), if the Sonos infrastructure is used with other services. If someone exclusively uses Music Assistant with Sonos speakers no problem. However throw into the mix playback from Spotify itself or in HA use of MMP to play something else breaks MA, as it can no longer playback.

MAMP = Music Assistant Media Player
MMP = Mini Media Player
HAMP = Home Assistant Media Player

Let's take the example of MMP and HAMP. I play something through the Windows Spotify app. Playback is properly displayed in the players.

![image](https://user-images.githubusercontent.com/87811684/229172951-79d54b09-6aee-43bf-9475-0e32d7e005cc.png)

However MA does not show what is currently been played, although the  queue seems correct. As you might notice from the image, playback is greyed out and not even restart of the MA add-on will bring it back live. Not certain if this is a known issue which will be fixed later. I am raising it now in case this has not been reported.

![image](https://user-images.githubusercontent.com/87811684/229173851-676410d1-80b6-495a-bbe8-9108805eab56.png)

The above issue is consistent on all my test HA instances, which are all setup with different options. MA playback is disabled on any Sonos asset and MA add-on restart, or even HA restart does not bring it back live.

This is what is shown concurrently in another instance

![image](https://user-images.githubusercontent.com/87811684/229175647-e2dac155-fd48-4b66-bb67-1ba36fe22a40.png)

This is another instance. As you can see things are inconsistent.

![image](https://user-images.githubusercontent.com/87811684/229178466-e84758f7-ea27-4561-94c7-dba3ad852dff.png)

One last thing it seems Airplay requires slimproto. I think you need to address this as people might disable slimproto, only to find out Airplay is broken. For newbies this might not be as easy to identify.

Thank you for a wonderful MA and please count me in on testing.

**Reply by marcelveldt** [reply]
Music Assistant focusses on being the non stop shop for streaming and account for other parties streaming to devices is not something I particular invested time in as its tricky to get that right cross-platform. So display the playback state is doable and maybe also the name of the source its playing (e.g. spotify) but anything else is probably out of scope for now at least.

yeah airplay for now relies on slimproto but that may change at some point. The interface should give a warning if you disable slimproto

**Reply by odowd** [reply]
Usual behaviour for an app is to just "grab" the Sonos speaker from whatever app is currently using it.  For example, My TV grabs my Sonos ARC when it is turned on, but I can play from Spotify on my phone, which silences the TV.  That's normal behaviour that MA should copy.  At the moment it is pretty useless because I cannot use MA after I have switched on my TV.  The MA-created music players work with oither Sonos players but not the ARC. It also refguses to work after I connect to the ARC and play music using the Sonos App (so it's not the TV that's the problem, apparently).

**Reply by marcelveldt** [reply]
I think we have a bug in that regard as someone else also mentioned that a sonos soundbar doesnt want to switch input - I'm investigating it.

**Comment by habitats-tech** [comment]
Yes I fully understand. The only functionality I am asking for is for MA to always show the current status of the player and artwork regardless which service was used to change its status. Also to be able to play anything at any time, regardless of what is currently being played on the specific player (no requirement to check if something is being played from wherever). Also I have no interest for MA to keep its queues in sync with other services. Focus only on the state of the player as every other HA media player behaves.

Hope the above makes sense without requiring heavy dev resources to address.

**Comment by midunker** [comment]
I can't use my Sonos Arc to play music. In the music assistant setting is an hourglass on the right side.
![image](https://github.com/music-assistant/hass-music-assistant/assets/97154661/e3aec3cc-a4a3-4657-bf9a-f02e707fe591)
If I choose this device nothing is played.
On the other hand, it works if I choose a chromcast player.
I try to use it in an automation like this:
```
service: media_player.play_media
target:
  entity_id: media_player.wohnzimmer
data:
  media_content_id: Eric Clapton
  media_content_type: music
```

**Reply by OzGav** [reply]
Did that player ever work?

**Reply by midunker** [reply]
Sure. I can play music with e.g. Spotcast or from the standard media lib.

**Reply by OzGav** [reply]
Sorry I meant did it ever work with MA?

**Reply by midunker** [reply]
I'm new to MA and installed it today. So no, it's a complete fresh installation.

**Reply by OzGav** [reply]
Try and reload the Sonos provider (in settings) and then look in the logs for any errors

**Reply by odowd** [reply]
Here is the warning in the log file:
2023-12-29 10:22:40.365 WARNING (MainThread) [music_assistant] Error while unload provider SONOS: UPnP Error 718 received: Invalid InstanceID from 192.168.1.152
2023-12-29 10:22:40.368 INFO (MainThread) [music_assistant] Loaded player provider SONOS

**Reply by OzGav** [reply]
Is that for a Sonos Arc?

**Reply by odowd** [reply]
Yes

**Reply by OzGav** [reply]
A user in a previous issue said

> I also had 'UPnP Error 718 received'
> I'm using Unifi AP and it startet working after activating Multicast DNS and IGMP Snooping in Global Network Settings

**Reply by odowd** [reply]
Thank you! I also have a Unifi router.  Multicast DNS was already on, but turning on IGMP Snooping solved the issue with the ARC.

**Comment by midunker** [comment]
Found this. Does it help?

2023-09-19 16:03:55.582 ERROR (MainThread) [music_assistant.webserver] [547344517840] Error handling message: CommandMessage(message_id=8, command='config/providers/reload', args={'instance_id': 'sonos'})
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/controllers/webserver.py", line 372, in _run_handler
    result = await result
             ^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/controllers/config.py", line 292, in reload_provider
    await self._load_provider_config(config)
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/controllers/config.py", line 699, in _load_provider_config
    await self.mass.load_provider(config)
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/server.py", line 359, in load_provider
    await self.unload_provider(conf.instance_id)
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/server.py", line 427, in unload_provider
    await provider.unload()
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/providers/sonos/__init__.py", line 258, in unload
    player.soco.end_direct_control_session()
  File "/usr/local/lib/python3.11/site-packages/soco/core.py", line 148, in inner_function
    return function(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/soco/core.py", line 846, in end_direct_control_session
    self.avTransport.EndDirectControlSession([("InstanceID", 0)])
  File "/usr/local/lib/python3.11/site-packages/soco/services.py", line 206, in _dispatcher
    return self.send_command(action, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/soco/services.py", line 514, in send_command
    self.handle_upnp_error(response.text)
  File "/usr/local/lib/python3.11/site-packages/soco/services.py", line 567, in handle_upnp_error
    raise SoCoUPnPException(
soco.exceptions.SoCoUPnPException: UPnP Error 718 received: Invalid InstanceID from 192.168.6.17

**Reply by midunker** [reply]
And this one a bit later

2023-09-19 16:05:43.765 ERROR (MainThread) [asyncio] Task exception was never retrieved
future: <Task finished name='Task-77965' coro=<SonosPlayerProvider._run_discovery() done, defined at /usr/local/lib/python3.11/site-packages/music_assistant/server/providers/sonos/__init__.py:414> exception=AttributeError("'NoneType' object has no attribute 'keys'")>
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/music_assistant/server/providers/sonos/__init__.py", line 427, in _run_discovery
    cur_player_ids = set(self.sonosplayers.keys())
                         ^^^^^^^^^^^^^^^^^^^^^^
AttributeError: 'NoneType' object has no attribute 'keys'

**Reply by OzGav** [reply]
Yeah ok. You will need to be patient until one of the devs can look at this. One of them uses an Arc so it should work but maybe a bug has been introduced that is causing the setup to fail. Your Chomecast option is the workaround for now.

**Reply by midunker** [reply]
ok, thx

**Reply by shapetheLOLa** [reply]
Having exactly the same with a Sonos port

**Reply by OzGav** [reply]
Guys if one of you can open an issue this will then not get lost and will get looked at in due course.

**Comment by phiebs** [comment]
Hi, is there any way to manually add Sonos players by their IP? Non-standard setup here but I managed to do it within HASS itself as autodiscover doesn't work for me (on purpose for various reasons) but unable to find a way to manually add Sonos players by IP in MASS?
Thanks in advance!

**Reply by marcelveldt** [reply]
No, we only support flat networks (so not routed vlans) OR simply make sure that HA is in your IoT vlan.
We have no plans of supporting other network setups as they always cause more pain than they do good.
When trying to use consumer grade equipment, setup your network for that and do not over-engineer it.

If you really want to isolate certain network equipment from your main network, just make sure HA operates in that network segment (you can even make HA multi homed if you want) and the same applies to MA ofcourse.

In the past years I've seen so many issues with multiple vLANS or even worse: mdns responders that I'm just rejecting any support as it caused to many issues for me to spend my (limited) time on.

**Reply by marcelveldt** [reply]
On a flat network where multicast mDNS discovery packets can travel freely, devices just work out of the box.

**Reply by phiebs** [reply]
Makes perfect sense, thank you for responding so quickly and descriptively.

**Reply by phiebs** [reply]
In case you were curious, I'm trying to "block" the kids from playing music on our Sonos equipment. Giving them their own VLAN for their devices, but still keeping control over their sonos's from a central place. Totally understand your reasoning, however!

**Reply by marcelveldt** [reply]
Interesting usecase. But my advise still stands: just add HA to the vLAN that contains the sonos and other devices. Easier to open a single HTTP(s) port to reach HA than trying to route all these protocols

**Reply by hansiputz** [reply]
Hi
I have the same config: HA and sonos in different networks. Routing would not be a problem - I can configure all that is needed - but I cant place them both in the same net. HA is in ioT net - Sonos in the same net as the clients and mobiles

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
